File: app/src/main/java/pl/srw/billcalculator/bill/calculation/TauronCalculatedBill.java
package pl.srw.billcalculator.bill.calculation;

import java.math.BigDecimal;

/**
 * Created by kseweryn on 09.06.15.
 */
public abstract class TauronCalculatedBill extends CalculatedEnergyBill {

    protected TauronCalculatedBill(String dateFrom, String dateTo, String oplataAbonamentowaPrice, String oplataPrzejsciowaPrice, String oplataStalaZaPrzesylPrice) {
        super(dateFrom, dateTo, oplataAbonamentowaPrice, oplataPrzejsciowaPrice, oplataStalaZaPrzesylPrice);
    }

    public abstract BigDecimal getSellNetCharge();

    public abstract BigDecimal getSellVatCharge();

    public final BigDecimal getSellGrossCharge() {
        return getSellNetCharge().add(getSellVatCharge());
    }

    public BigDecimal getDistributeNetCharge() {
        return getNetChargeSum().subtract(getSellNetCharge());
    }

    public BigDecimal getDistributeVatCharge() {
        return getVatChargeSum().subtract(getSellVatCharge());
    }

    public final BigDecimal getDistributeGrossCharge() {
        return getDistributeNetCharge().add(getDistributeVatCharge());
    }
}

File: app/src/main/java/pl/srw/billcalculator/bill/calculation/TauronCalculatedBill.kt
package pl.srw.billcalculator.bill.calculation

import org.threeten.bp.LocalDate
import pl.srw.billcalculator.db.Prices
import java.math.BigDecimal

abstract class TauronCalculatedBill(
    dateFrom: LocalDate,
    dateTo: LocalDate,
    oplataAbonamentowaPrice: String,
    oplataPrzejsciowaPrice: String,
    oplataStalaZaPrzesylPrice: String,
    prices: Prices
) : CalculatedEnergyBill(
    dateFrom,
    dateTo,
    oplataAbonamentowaPrice,
    oplataPrzejsciowaPrice,
    oplataStalaZaPrzesylPrice,
    prices
) {

    abstract val sellNetCharge: BigDecimal

    abstract val sellVatCharge: BigDecimal

    val sellGrossCharge: BigDecimal
        get() = sellNetCharge.add(sellVatCharge)

    val distributeNetCharge: BigDecimal
        get() = netChargeSum.subtract(sellNetCharge)

    val distributeVatCharge: BigDecimal
        get() = vatChargeSum.subtract(sellVatCharge)

    val distributeGrossCharge: BigDecimal
        get() = distributeNetCharge.add(distributeVatCharge)
}
